File: app/src/main/java/eu/warble/pjapp/ui/map/MapListFragment.java
package eu.warble.pjapp.ui.map;


import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ListView;
import android.widget.ProgressBar;

import com.indoorway.android.common.sdk.model.IndoorwayBuilding;

import java.util.List;

import eu.warble.pjapp.R;
import eu.warble.pjapp.data.model.MapListItem;
import eu.warble.pjapp.ui.base.BaseFragment;

public class MapListFragment extends BaseFragment<MapListPresenter> {

    private ListView mapList;
    private ProgressBar progressBar;

    @Override
    protected MapListPresenter createPresenter() {
        return new MapListPresenter(this);
    }

    public static MapListFragment newInstance() {
        Bundle args = new Bundle();
        MapListFragment fragment = new MapListFragment();
        fragment.setArguments(args);
        fragment.tag = "MapListFragment";
        return fragment;
    }

    @Override
    protected View initView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_map_list, container, false);
        setHasOptionsMenu(true);
        initViews(view);
        return view;
    }

    private void initViews(View view) {
        MapListAdapter adapter = new MapListAdapter(context.getApplicationContext());
        progressBar = view.findViewById(R.id.progressBar);
        mapList = view.findViewById(R.id.list_maps);
        mapList.setAdapter(adapter);
        mapList.setOnItemClickListener((parent, view1, position, id) -> {
            MapListItem item = adapter.getItem(position);
            presenter.onMapItemClicked(item);
        });
    }

    public void updateList(List<IndoorwayBuilding> newData) {
        MapListAdapter adapter = (MapListAdapter) mapList.getAdapter();
        adapter.updateList(newData);
    }

    void setLoadingState(boolean state){
        mapList.setVisibility(state ? View.GONE : View.VISIBLE);
        progressBar.setVisibility(state ? View.VISIBLE : View.GONE);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        inflater.inflate(R.menu.toolbar, menu);
    }

    void startMapActivity(Intent intent){
        startActivity(intent);
    }
}

File: app/src/main/java/eu/warble/pjappkotlin/mvp/map/list/MapListFragment.kt
package eu.warble.pjappkotlin.mvp.map.list

import android.content.Context
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.indoorway.android.common.sdk.model.IndoorwayBuilding
import com.indoorway.android.common.sdk.model.IndoorwayObjectId
import eu.warble.pjappkotlin.R
import eu.warble.pjappkotlin.mvp.ApplicationNavigator
import eu.warble.pjappkotlin.mvp.BaseActivity
import eu.warble.pjappkotlin.mvp.BaseFragment
import eu.warble.pjappkotlin.utils.Constants.EMPTY_STRING
import kotlinx.android.synthetic.main.fragment_map_list.view.btn_current_location
import kotlinx.android.synthetic.main.fragment_map_list.view.items_list
import kotlinx.android.synthetic.main.fragment_map_list.view.loading_screen

class MapListFragment : BaseFragment(), MapListContact.View {

    override val TAG: String = "MapListFragment"

    override lateinit var presenter: MapListContact.Presenter
    override val applicationNavigator: ApplicationNavigator by lazy {
        ApplicationNavigator(activity as BaseActivity)
    }
    private lateinit var itemsList: RecyclerView
    private lateinit var locationFAB: FloatingActionButton
    private lateinit var loadingScreen: View
    private lateinit var buildingsAdapter: BuildingsListAdapter
    private lateinit var mapsAdapter: MapsListAdapter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_map_list, container, false)
        initViews(view)
        initAdapter()
        initListener()
        initPresenter()
        return view
    }

    private fun initViews(view: View) {
        itemsList = view.items_list
        locationFAB = view.btn_current_location
        loadingScreen = view.loading_screen
    }

    private fun initListener() {
        locationFAB.setOnClickListener {
            findMyLocation()
        }
    }

    private fun initAdapter() {
        buildingsAdapter = BuildingsListAdapter(emptyList(), { buildingItem: IndoorwayBuilding ->
            presenter.loadMaps(buildingItem)
        })
        mapsAdapter = MapsListAdapter(emptyList(), { mapItem: IndoorwayObjectId ->
            showMap(presenter.selectedBuilding?.uuid ?: EMPTY_STRING, mapItem)
        })
    }

    override fun updateMapsListItems(items: List<IndoorwayObjectId>) {
        mapsAdapter.updateList(items)
        itemsList.adapter = mapsAdapter
        itemsList.layoutManager = LinearLayoutManager(mContext)
        itemsList.invalidate()
    }

    override fun updateBuildingsListItems(items: List<IndoorwayBuilding>) {
        buildingsAdapter.updateList(items)
        itemsList.adapter = buildingsAdapter
        itemsList.layoutManager = GridLayoutManager(mContext, 2)
        itemsList.invalidate()
    }

    override fun showMap(buildingUUID: String, map: IndoorwayObjectId) {
        applicationNavigator.goToMapActivity(buildingUUID, map)
    }

    private fun findMyLocation() {
        applicationNavigator.goToMapActivity()
    }

    override fun showLoadingScreen(show: Boolean) {
        loadingScreen.visibility = if (show) View.VISIBLE else View.GONE
        locationFAB.visibility = if (show) View.GONE else View.VISIBLE
    }

    override fun onBack(): Boolean {
        return when (itemsList.adapter) {
            is MapsListAdapter -> {
                presenter.loadBuildings()
                true
            }
            else -> false
        }
    }

    private fun initPresenter() {
        presenter = MapListPresenter(
                this,
                mContext as Context
        )
        presenter.start()
    }

    companion object {

        fun newInstance() = MapListFragment()

    }
}